Feature/chupep/lepl 16

Samael1414 opened the pull request:
Создана таблица с перечнем тэгов для курсвов. Создан скрипт миграции с инсертом тэгов. Связаны сущность тэгов и курсов через промежуточную таблицу. В респонс курсов добавлен показ тэгов. При создании курса есть возможность выбрать тэг. Создан отдельный эндпоинт в api-gateway с редиректорм по gRPC. Добавлены тесты на контроллер по управлению тэгами, и метод созданию курсов. Добавлена javadoc

VladislavYurin commented on api-gateway/src/main/java/ru/mentor/controller/TagController.java:
```diff
@@ -0,0 +1,55 @@
+ * и возвращает результат в виде DTO-моделей {@link TagDto}.
+ * </p>
+ *
+ * <p><b>Эндпоинты:</b></p>
```
лишнее перечисление методов

VladislavYurin commented on api-gateway/src/main/java/ru/mentor/controller/TagController.java:
```diff
@@ -0,0 +1,55 @@
+    @GetMapping("/all")
+    public ResponseEntity<List<TagDto>> getAll() {
+        ListTagsResponse response = client.listTags();
+        List<TagDto> list = response.getTagsList().stream()
```
делай редирект сервис в который выносим вызов клиента + логирование нужно и обработку ошибок

VladislavYurin commented on api-gateway/src/main/java/ru/mentor/mapper/ProtoTimeMapper.java:
```diff
@@ -0,0 +1,40 @@
+ */
+@Component
+@RequiredArgsConstructor
+public class ProtoTimeMapper {
```
а если потом в будущем у нас будет грпс общение не только между апи-гейтвеем, а допустим между календарем и курс-сервисом (например), то этот маппер не будет доступен. Надо в common-lib

VladislavYurin commented on common-lib/src/main/java/ru/mentor/entity/CourseTagId.java:
```diff
@@ -0,0 +1,28 @@
+
+@NoArgsConstructor
+@AllArgsConstructor
+public class CourseTagId implements Serializable {
```
а зачем это вообще нужно для жавы?

Samael1414 committed:
Создана таблица с перечнем тэгов для курсвов. Создан скрипт миграции с
инсертом тэгов. Связаны сущность тэгов и курсов через промежуточную
таблицу. В респонс курсов добавлен показ тэгов. При создании курса есть
возможность выбрать тэг. Создан отдельный эндпоинт в api-gateway с
редиректорм по gRPC. Добавлены тесты на контроллер по управлению тэгами,
и метод созданию курсов. Добавлена javaDoc. Добавлены тесты и эндпойнты в Insomnia_endpoints.yaml

VladislavYurin reviewed (commented):
javaDoc везде где нет
зачем нужен tagController в course-service если он нигде не вызывается?

Samael1414 commented:
> javaDoc везде где нет
> 
> зачем нужен tagController в course-service если он нигде не вызывается?
> 
> 

А разве не нужна история по управлению тэгами? Так ты писал убрать javaDoc разве нет ?

VladislavYurin commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -73,19 +67,12 @@ public CourseDto createCourse(InnerCreateCourseRequest request) {
-     * @throws CustomAccessDeniedException Если пользователь не имеет прав для удаления курса.
-     */
     @Override
+    @Transactional
```
а зачем транзакция если у нас атомарные операции здесь

VladislavYurin replied:
!

VladislavYurin commented on common-lib/src/main/java/ru/mentor/repository/CourseRepository.java:
```diff
@@ -1,49 +1,26 @@
 import ru.mentor.entity.CourseEntity;
 import ru.mentor.exception.EntityNotFoundException;
 
-/**
```
зачем весь жавадок удалили

VladislavYurin commented on api-gateway/pom.xml:
```diff
@@ -77,5 +77,15 @@
       <version>2.2.30</version>
     </dependency>
+    <dependency>
+      <groupId>org.springframework.security</groupId>
```
а до этого они где были?

VladislavYurin commented on api-gateway/src/main/java/ru/mentor/controller/TagController.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/tag")
```
у нас будет несколько типов тэгов, для курсов и для менторов, надо их разбить по неймингу и в дто и в ендпоинтах

VladislavYurin commented on api-gateway/src/main/java/ru/mentor/grpc/TagsGrpcClient.java:
```diff
@@ -0,0 +1,22 @@
+
+@Component
+@NoArgsConstructor
+@AllArgsConstructor
```
requiredArgsConstructor, остальные не нужны

VladislavYurin commented on api-gateway/src/main/java/ru/mentor/services/impl/RedirectTagServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    public ResponseEntity<List<TagDto>> getAllTags() {
+        String rqId = RqGenerator.generateRqId();
+        log.info("[RqUid = {} ] Запрос на получение списка тегов", rqId);
```
каких тегов?

VladislavYurin commented on api-gateway/src/main/java/ru/mentor/services/impl/RedirectTagServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+            return ResponseEntity.ok(list);
+        } catch (Exception e) {
+            log.error("[ RqUId = {} ] Ошибка при получении списка тегов", rqId, e);
+            return ResponseEntity.internalServerError().build();
```
то есть всегда срем статус 500 клиенту в лицо даже без сообщения?

VladislavYurin commented on api-gateway/src/main/resources/db.changelog/sql/init_tables.sql:
```diff
@@ -91,6 +91,29 @@ CREATE TABLE IF NOT EXISTS mentor_time_slot__users
     user_id      BIGINT NOT NULL
 );
 
+CREATE TABLE IF NOT EXISTS tags
```
переименовать бд, теги будут и для курсов и для менторов, по наименованию не понятно к чему они относятся

VladislavYurin commented on api-gateway/src/main/resources/db.changelog/sql/insert_init_content.sql:
```diff
@@ -43,4 +44,39 @@ $$),
+$$);
+
+
+INSERT INTO tags (tag_name)
```
перевести этот скрипт в init_tables, этот скрипт для тестовых данных

VladislavYurin commented on common-lib/pom.xml:
```diff
@@ -191,6 +213,33 @@
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.11.0</version>
+        <configuration>
+          <source>17</source>
```
уже 21

VladislavYurin commented on course-service/src/main/java/ru/mentor/grpc/TagsGrpcServer.java:
```diff
@@ -0,0 +1,54 @@
+
+    private final TagRepository tagRepository;
+
+    public TagsGrpcServer(@Qualifier("tagMapper") TagMapper mapper, TagRepository tagRepository) {
```
почему не аннотация?

VladislavYurin commented on course-service/src/main/java/ru/mentor/grpc/TagsGrpcServer.java:
```diff
@@ -0,0 +1,54 @@
+                .stream()
+                .map(mapper::toDto)
+                .toList();
+        for (TagDto tag : tags) {
```
стрим

VladislavYurin commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -1,35 +1,24 @@
-import ru.mentor.repository.UserCourseAccessRepository;
-import ru.mentor.repository.UserModuleAccessRepository;
-import ru.mentor.repository.UserRepository;
+import ru.mentor.repository.*;
```
за * ору в импортах

jammyHubby999 replied:
все исправм, все уберем)

VladislavYurin commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -1,35 +1,24 @@
- */
 @Service
 @RequiredArgsConstructor
+@Transactional(readOnly = true)
```
на демо сможете пояснить зачем транзакция?

jammyHubby999 replied:
я не спец по транзакциям, но насколько я понимаю @Transactional(readOnly = true) здесь, потому что в классе есть get-методы, которые внутри себя делают несколько вызовов репозиториев.  Если нет внешней @Transactional, каждый вызов JPA репозитория создает отдельную транзакцию, что не хорошо. readOnly = true определено на уровне класса, но подойдет только для get методов. Над методами create и delete нужно переопределить @Transactional  на read-write.

VladislavYurin commented on course-service/src/test/java/ru/mentor/serviceImpl/CourseServiceImplTest.java:
```diff
@@ -0,0 +1,176 @@
+    CourseServiceImpl service;
+
+    @Nested
+    class CreateCourse {
```
это что за чудо конструкция, почему сразу нельзя писать метода теста?

jammyHubby999 replied:
Это и првда чудо-конструкция. Вообще по сути оно бывает полезно, когда есть несколько тестов, которые относятся к одному методу, например create - на него несколько тестов, delete - на него еще несколько. Так их группируют - в один @Nested пишут все тесты для create метода, в другой @Nested все для delete и т. д. Но здесь это явно избыточно сделали.

muzkatttt reviewed (commented):
отличная работа!

VladislavYurin commented on api-gateway/src/main/resources/db.changelog/sql/init_tables.sql:
```diff
@@ -91,6 +91,29 @@ CREATE TABLE IF NOT EXISTS mentor_time_slot__users
     user_id      BIGINT NOT NULL
 );
 
+CREATE TABLE IF NOT EXISTS course_tags
```
вынести в отдельный скрипт и отдельный ченжсет

povitoi commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -40,29 +39,39 @@ public class CourseServiceImpl implements CourseService {
-
-        // Проверяем, является ли пользователь админом или ментором
         if (Role.checkIsMentor(user) || Role.checkIsAdmin(user)) {
             CourseEntity course = new CourseEntity();
```
У этой энтити билдер же есть)

povitoi commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -40,29 +39,39 @@ public class CourseServiceImpl implements CourseService {
             course.setAuthor(user);
             course.setCourseTitle(request.getCourseName());
             course.setDescription(request.getCourseDescription());
+            List<Long> ids = request.getTagIds() == null ? List.of() : request.getTagIds().stream().distinct().toList();
```
Я бы вынес вот этот кусок в конце в мини метод, который имел бы красивое название, типа вернутьКрутойСписок и можно там его вызывать. Будет намного понятнее что там происходит просто по названию метода. А сейчас сложная цепочка вызовов

povitoi commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -40,29 +39,39 @@ public class CourseServiceImpl implements CourseService {
         UserEntity user = userRepository.findByIdOrThrow(request.getAuthorId());
-
-        // Проверяем, является ли пользователь админом или ментором
         if (Role.checkIsMentor(user) || Role.checkIsAdmin(user)) {
```
Почему здесь не используются hasRole из спринг секьюрити? Они же буквально предназначены для этого

povitoi commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -74,18 +83,18 @@ public CourseDto createCourse(InnerCreateCourseRequest request) {
 
-        // Проверка на права доступа:
         // Админ может удалять любой курс
         if (Role.checkIsAdmin(deletedByUser)) {
```
Тот же hasRole

povitoi commented on course-service/src/main/java/ru/mentor/service/impl/CourseServiceImpl.java:
```diff
@@ -109,112 +118,88 @@ public void deleteCourse(Long userId, Long courseId) {
         UserEntity user = userRepository.findByIdOrThrow(userId);
-
-        // Админ может получить все активные курсы
         if (Role.checkIsAdmin(user)) {
```
hasRole

povitoi replied:
Надо везде проверить)

povitoi commented on course-service/src/test/java/ru/mentor/serviceImpl/CourseServiceImplTest.java:
```diff
@@ -0,0 +1,172 @@
+@ExtendWith(MockitoExtension.class)
+public class CourseServiceImplTest {
+
+    private static final Long AUTHOR_ID = 100L;
```
У нас в коммон либе есть классы, которые содержат тестовые данные вот такие

povitoi commented on course-service/src/test/java/ru/mentor/serviceImpl/CourseServiceImplTest.java:
```diff
@@ -0,0 +1,172 @@
+            mockedRole.when(() -> Role.checkIsMentor(user)).thenReturn(true);
+            mockedRole.when(() -> Role.checkIsAdmin(user)).thenReturn(false);
+
+            CourseTagEntity tagOne = new CourseTagEntity();
```
Все вот такие моменты лучше перенести в те классы с тестовыми данными
